File: src/api/schemas/providers.py
"""
Provider and Model API Schemas.

Pydantic models for LLM provider and model management endpoints.
"""

from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field

# ============================================================================
# Provider Schemas
# ============================================================================


class ProviderBase(BaseModel):
    """Base provider fields."""

    provider_id: str = Field(
        ..., min_length=1, max_length=50, description="Unique provider identifier"
    )
    display_name: str = Field(..., min_length=1, max_length=100, description="Display name")
    api_type: str = Field(
        default="openai",
        description="API type: openai, anthropic, google, or google-vertex",
    )
    base_url: str | None = Field(None, max_length=500, description="Custom API base URL")
    metadata: dict[str, Any] = Field(default_factory=dict, description="Safe provider metadata")
    is_enabled: bool = Field(default=True, description="Whether the provider is enabled")


class ProviderCreate(ProviderBase):
    """Request to create a provider."""

    api_key: str | None = Field(None, description="API key (will be encrypted)")


class ProviderUpdate(BaseModel):
    """Request to update a provider."""

    display_name: str | None = Field(None, min_length=1, max_length=100)
    api_type: str | None = None
    base_url: str | None = Field(None, max_length=500)
    metadata: dict[str, Any] | None = Field(None, description="Safe provider metadata")
    api_key: str | None = Field(None, description="New API key (will be encrypted)")
    is_enabled: bool | None = None


class ProviderResponse(ProviderBase):
    """Provider response (without API key)."""

    tenant_id: str
    has_api_key: bool = Field(description="Whether API key is configured")
    allow_environment_credentials: bool = Field(
        default=False,
        description="Whether the provider can use server-side environment credentials",
    )
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ProviderListResponse(BaseModel):
    """List of providers response."""

    providers: list[ProviderResponse]
    total: int


class ProviderTestResult(BaseModel):
    """Result of testing provider connection."""

    success: bool
    message: str
    latency_ms: int | None = None


class ProviderTemplateCredentialField(BaseModel):
    """Credential field exposed by a provider template."""

    name: str
    label: str
    field_type: str = "password"
    required: bool = True
    placeholder: str | None = None


class ProviderTemplateModel(BaseModel):
    """Trusted model metadata exposed by a provider template."""

    model_id: str
    display_name: str
    context_window: int
    max_output_tokens: int
    supports_vision: bool
    supports_tools: bool
    input_price_per_1k: float
    output_price_per_1k: float
    access_level: str
    sort_order: int


class ProviderTemplateResponse(BaseModel):
    """Provider template response for guided onboarding."""

    template_id: str
    display_name: str
    description: str
    default_provider_id: str
    api_type: str
    default_base_url: str
    credential_fields: list[ProviderTemplateCredentialField]
    discovery_strategy: str
    default_models: list[ProviderTemplateModel]
    advanced: bool = False


class ProviderFromTemplateCreate(BaseModel):
    """Request to create or update a provider from a template."""

    template_id: str = Field(..., min_length=1)
    provider_id: str | None = Field(None, min_length=1, max_length=50)
    display_name: str | None = Field(None, min_length=1, max_length=100)
    base_url: str | None = Field(None, max_length=500)
    metadata: dict[str, Any] | None = Field(None, description="Safe provider metadata")
    api_key: str | None = Field(None, description="API key (will be encrypted)")
    is_enabled: bool = True


class ProviderModelSyncSkipped(BaseModel):
    """A discovered model that was not enabled by sync."""

    model_id: str
    reason: str


class ProviderModelSyncItem(BaseModel):
    """A model created or updated by provider model sync."""

    model_id: str
    provider_id: str
    display_name: str
    is_enabled: bool


class ProviderModelSyncResult(BaseModel):
    """Provider model sync response."""

    provider_id: str
    template_id: str | None = None
    created_models: list[ProviderModelSyncItem]
    updated_models: list[ProviderModelSyncItem]
    skipped_models: list[ProviderModelSyncSkipped]
    discovery_warnings: list[str]


# ============================================================================
# Model Schemas
# ============================================================================


class ModelBase(BaseModel):
    """Base model fields."""

    model_id: str = Field(
        ..., min_length=1, max_length=100, description="Model identifier for API calls"
    )
    provider_id: str = Field(
        ..., min_length=1, max_length=50, description="Provider this model belongs to"
    )
    display_name: str = Field(..., min_length=1, max_length=100, description="Display name")
    context_window: int = Field(default=128000, ge=1, description="Context window size in tokens")
    max_output_tokens: int = Field(default=4096, ge=1, description="Maximum output tokens")
    supports_vision: bool = Field(
        default=False, description="Whether the model supports vision/images"
    )
    supports_tools: bool = Field(
        default=True, description="Whether the model supports tool calling"
    )
    input_price_per_1k: Decimal = Field(
        default=Decimal("0"), ge=0, description="Input price per 1K tokens in USD"
    )
    output_price_per_1k: Decimal = Field(
        default=Decimal("0"), ge=0, description="Output price per 1K tokens in USD"
    )
    access_level: str = Field(
        default="public", description="Access level: public, premium, or admin"
    )
    is_enabled: bool = Field(default=True, description="Whether the model is enabled")
    sort_order: int = Field(default=0, description="Sort order for UI display")


class ModelCreate(ModelBase):
    """Request to create a model."""

    pass


class ModelUpdate(BaseModel):
    """Request to update a model.

    ``model_id`` is renameable — the frontend now exposes it in the edit
    dialog so a provider's model rename (e.g. ``gemini-3-flash-preview`` →
    ``gemini-3.1-flash``) doesn't require delete+recreate. Server-side
    rename pushes an UPDATE on the primary key; the pricing-table sync
    also follows the new id.
    """

    model_id: str | None = Field(None, min_length=1, max_length=100)
    display_name: str | None = Field(None, min_length=1, max_length=100)
    context_window: int | None = Field(None, ge=1)
    max_output_tokens: int | None = Field(None, ge=1)
    supports_vision: bool | None = None
    supports_tools: bool | None = None
    input_price_per_1k: Decimal | None = Field(None, ge=0)
    output_price_per_1k: Decimal | None = Field(None, ge=0)
    access_level: str | None = None
    is_enabled: bool | None = None
    sort_order: int | None = None


class ModelResponse(ModelBase):
    """Model response."""

    tenant_id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ModelListResponse(BaseModel):
    """List of models response."""

    models: list[ModelResponse]
    total: int


class ModelToggleRequest(BaseModel):
    """Request to toggle model enabled state."""

    is_enabled: bool

File: web/src/api/providers.ts
/**
 * LLM Providers API client.
 *
 * Provides CRUD operations for managing LLM providers.
 */

import { api } from "@/lib/api";

// =============================================================================
// Types
// =============================================================================

export interface Provider {
  provider_id: string;
  tenant_id: string;
  display_name: string;
  api_type: string;
  base_url?: string | null;
  has_api_key: boolean;
  is_enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface ProviderCreate {
  provider_id: string;
  display_name: string;
  api_type?: string;
  base_url?: string;
  api_key?: string;
  is_enabled?: boolean;
}

export interface ProviderTemplateCredentialField {
  name: string;
  label: string;
  field_type: "password" | "text" | string;
  required: boolean;
  placeholder?: string | null;
}

export interface ProviderTemplateModel {
  model_id: string;
  display_name: string;
  context_window: number;
  max_output_tokens: number;
  supports_vision: boolean;
  supports_tools: boolean;
  input_price_per_1k: number;
  output_price_per_1k: number;
  access_level: string;
  sort_order: number;
}

export interface ProviderTemplate {
  template_id: string;
  display_name: string;
  description: string;
  default_provider_id: string;
  api_type: string;
  default_base_url: string;
  credential_fields: ProviderTemplateCredentialField[];
  discovery_strategy: string;
  default_models: ProviderTemplateModel[];
  advanced: boolean;
}

export interface ProviderFromTemplateCreate {
  template_id: string;
  provider_id?: string;
  display_name?: string;
  api_key?: string;
  base_url?: string;
  is_enabled?: boolean;
}

export interface ProviderUpdate {
  display_name?: string;
  api_type?: string;
  base_url?: string;
  api_key?: string;
  is_enabled?: boolean;
}

export interface ProviderTestResult {
  success: boolean;
  message: string;
  latency_ms?: number;
}

export interface ProviderModelSyncItem {
  model_id: string;
  provider_id: string;
  display_name: string;
  is_enabled: boolean;
}

export interface ProviderModelSyncSkipped {
  model_id: string;
  reason: string;
}

export interface ProviderModelSyncResult {
  provider_id: string;
  template_id?: string | null;
  created_models: ProviderModelSyncItem[];
  updated_models: ProviderModelSyncItem[];
  skipped_models: ProviderModelSyncSkipped[];
  discovery_warnings: string[];
}

export const providerQueryKeys = {
  all: ["providers"] as const,
  list: (includeDisabled = false) => ["providers", { includeDisabled }] as const,
  templates: ["provider-templates"] as const,
};

// =============================================================================
// API Functions
// =============================================================================

/**
 * List all providers.
 */
export async function listProviders(includeDisabled = false): Promise<Provider[]> {
  const { data } = await api.get<Provider[]>(
    `/api/v1/providers?include_disabled=${includeDisabled}`
  );
  return data;
}

/**
 * List guided provider templates.
 */
export async function listProviderTemplates(): Promise<ProviderTemplate[]> {
  const { data } = await api.get<ProviderTemplate[]>("/api/v1/provider-templates");
  return data;
}

/**
 * Get a specific provider.
 */
export async function getProvider(providerId: string): Promise<Provider> {
  const { data } = await api.get<Provider>(`/api/v1/providers/${providerId}`);
  return data;
}

/**
 * Create a new provider.
 */
export async function createProvider(provider: ProviderCreate): Promise<Provider> {
  const { data } = await api.post<Provider>("/api/v1/providers", provider);
  return data;
}

/**
 * Create a provider from a guided template.
 */
export async function createProviderFromTemplate(
  provider: ProviderFromTemplateCreate
): Promise<Provider> {
  const { data } = await api.post<Provider>("/api/v1/providers/from-template", provider);
  return data;
}

/**
 * Update a provider.
 */
export async function updateProvider(
  providerId: string,
  updates: ProviderUpdate
): Promise<Provider> {
  const { data } = await api.put<Provider>(`/api/v1/providers/${providerId}`, updates);
  return data;
}

/**
 * Delete a provider.
 */
export async function deleteProvider(providerId: string): Promise<void> {
  await api.delete(`/api/v1/providers/${providerId}`);
}

/**
 * Test provider API connection.
 */
export async function testProviderConnection(providerId: string): Promise<ProviderTestResult> {
  const { data } = await api.post<ProviderTestResult>(
    `/api/v1/providers/${providerId}/test`
  );
  return data;
}

/**
 * Sync provider-supported models from the Gateway catalog/discovery layer.
 */
export async function syncProviderModels(providerId: string): Promise<ProviderModelSyncResult> {
  const { data } = await api.post<ProviderModelSyncResult>(
    `/api/v1/providers/${providerId}/models/sync`
  );
  return data;
}

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Get API type display name.
 */
export function getApiTypeDisplayName(apiType: string): string {
  const names: Record<string, string> = {
    openai: "OpenAI Compatible",
    dashscope: "DashScope",
    anthropic: "Anthropic",
    google: "Google Gemini",
    "google-vertex": "Google Vertex AI",
  };
  return names[apiType] || apiType;
}

/**
 * Get default base URL for API type.
 */
export function getDefaultBaseUrl(apiType: string): string {
  const urls: Record<string, string> = {
    openai: "https://api.openai.com",
    dashscope: "https://dashscope.aliyuncs.com/compatible-mode",
    anthropic: "https://api.anthropic.com",
    google: "https://generativelanguage.googleapis.com",
    "google-vertex": "https://aiplatform.googleapis.com",
  };
  return urls[apiType] || "";
}
